flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/utils/FlinkResourceExceptionUtils.java

```diff
@@ -0,0 +1,118 @@
+import org.apache.flink.kubernetes.operator.exception.DeploymentFailedException;
+import org.apache.flink.kubernetes.operator.exception.FlinkResourceException;
+import org.apache.flink.kubernetes.operator.exception.ReconciliationException;
+import org.apache.flink.runtime.rest.util.RestClientException;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/** Flink Resource Exception utilities. */
+public final class FlinkResourceExceptionUtils {
+
+    public static <R extends AbstractFlinkResource> void updateFlinkResourceException(
+            Throwable throwable, R resource) {
+
+        FlinkResourceException flinkResourceException =
+                getFlinkResourceException(throwable, getStackTraceEnabled(resource));
+
+        try {
+            ((AbstractFlinkResource<?, ?>) resource)
+                    .getStatus()
+                    .setError(convertToJson(flinkResourceException));
+        } catch (Exception e) {
+            // Rollback to setting error string/message to CRD
+            ((AbstractFlinkResource<?, ?>) resource)
+                    .getStatus()
+                    .setError(
+                            (e instanceof ReconciliationException)
+                                    ? e.getCause().toString()
+                                    : e.toString());
+        }
+    }
+
+    private static <R extends AbstractFlinkResource> boolean getStackTraceEnabled(R resource) {
+        return Optional.ofNullable((AbstractFlinkSpec) resource.getSpec())
+                .map(AbstractFlinkSpec::isStackTraceEnabled)
+                .orElse(false);
+    }
+
+    private static FlinkResourceException getFlinkResourceException(
+            Throwable throwable, boolean isStackTraceEnabled) {
+        FlinkResourceException flinkResourceException =
+                convertToFlinkResourceException(throwable, isStackTraceEnabled);
+
+        flinkResourceException.setThrowableList(
+                ExceptionUtils.getThrowableList(
+                                throwable.getCause())
+                        .stream()
+                        .map((t) -> convertToFlinkResourceException(t, false))
+                        .collect(Collectors.toList()));
+
+        return flinkResourceException;
+    }
+
+    private static FlinkResourceException convertToFlinkResourceException(
+            Throwable throwable, boolean isStackTraceEnabled) {
+        FlinkResourceException flinkResourceException =
+                FlinkResourceException.builder()
+                        .type(throwable.getClass().getName())
+                        .message(throwable.getMessage())
+                        .build();
+
+        if (isStackTraceEnabled) {
+            flinkResourceException.setStackTraceElements(throwable.getStackTrace());
```

Comment:
We should have size limits on the stack trace as well, how many lines, total length etc.

Reply:
Added ConfigOption to limit stackTrace list size